File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/map/TileLodMode.java
package org.maplibre.nativeffi.map;

/** Tile level-of-detail algorithm used by map tile options. */
public final class TileLodMode {
  public static final TileLodMode DEFAULT = new TileLodMode(0);
  public static final TileLodMode DISTANCE = new TileLodMode(1);

  private final int rawValue;
  private final String name;

  public TileLodMode(int rawValue) {
    this.rawValue = rawValue;
    this.name =
        switch (rawValue) {
          case 0 -> "DEFAULT";
          case 1 -> "DISTANCE";
          default -> "UNKNOWN(" + Integer.toUnsignedLong(rawValue) + ")";
        };
  }

  public int rawValue() {
    return rawValue;
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof TileLodMode that && rawValue == that.rawValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(rawValue);
  }

  @Override
  public String toString() {
    return name;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/map/TileLodMode.kt
package org.maplibre.nativeffi.map

import kotlin.jvm.JvmInline

/**
 * Tile level-of-detail algorithm used by map tile options.
 *
 * This is an open domain: MapLibre Native may report a value that has no named constant here, so a
 * `when` over this type needs an `else` branch. Unknown values are preserved as their raw
 * [nativeValue] rather than collapsed to a known constant.
 */
@JvmInline
public value class TileLodMode(public val nativeValue: Int) {
  public companion object {
    public val DEFAULT: TileLodMode = TileLodMode(0)
    public val DISTANCE: TileLodMode = TileLodMode(1)

    internal fun fromNative(nativeValue: UInt): TileLodMode = fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): TileLodMode = TileLodMode(nativeValue)
  }

  internal val isKnown: Boolean
    get() = this == DEFAULT || this == DISTANCE
}
